### PR TITLE
Don't NCPAINT our window, PAINT our window

### LINKED ISSUE #872: We need to draw the whole non-client area ourselves
If we want to be able to fully theme the window, we'll probably need to draw the theme parts ourselves. Right now the close button, minimize button and maximize/restore buttons are drawn by DWM/uxtheme if we don't tell it not to.
We don't get much say in where they are displayed (which is why there's a border on the right side of the window), and if we attempt to draw behind them we get an additive blend when the DWM splats the title bar down on top of us.

We probably need to look at [`DrawThemeBackground`](https://docs.microsoft.com/en-us/windows/desktop/api/uxtheme/nf-uxtheme-drawthemebackground), [`DrawThemeIcon`](https://docs.microsoft.com/en-us/windows/desktop/api/uxtheme/nf-uxtheme-drawthemeicon) and friends.

### PR DESCRIPTION
**Pull request by zadjii-msft**
## Summary of the Pull Request

Fixes the round corners and lack of shadow on the window introduced in #929.
image

 
## References
This introduces the problem discussed in #1897

## PR Checklist
* [x] Touches some #1625 bits. Closes #872 
* [x] you better believe I work here
* [ ] Tests make me cry
* [x] Requires documentation to be updated - no

## Detailed Description of the Pull Request / Additional comments

* If we handle `WM_NCPAINT`, we _don't_ get shadows. We also get the legacy rounded top window corners, because of some kernel insanity. The whole point of `DwmExtendFrameIntoClientArea` is that we can just `PAINT` the entire window, instead of trying to do `NC_PAINT` tricks. So, we need to move the entire NCPAINT we had into PAINT.
* We also need to remove the margins we were setting. If we leave a top margin, the system caption buttons will try and draw into that area. Since we don't want those, we need to make sure the top margin is 0, so they have no space to draw into. 
* I also split out the DragSizeChanged handler from OnSize. I don't know why, but having it inside OnSize would cause the xaml island content to go black when moving the window from a secondary normal DPI display back to a high DPI primary display.

### CODE REVIEW

**Comment by DHowett-MSFT** [comment]
Where are the resize handles here? Are they _outside, on the shadow_ or _inside, in the white border_?

**Comment by zadjii-msft** [comment]
> Where are the resize handles here? Are they _outside, on the shadow_ or _inside, in the white border_?

They are exactly on the border, unfortunately. No _worse_ than before. I might be able to play with our NCHITTEST code - maybe **we're**  clipping the resize handle area, though no promises.

File: src/cascadia/WindowsTerminal/NonClientIslandWindow.cpp [hunk 1]
```diff
@@ -113,8 +118,53 @@ void NonClientIslandWindow::OnSize(const UINT width, const UINT height)
         _rootGrid.Arrange(finalRect);
     }
 
+    winrt::check_bool(SetWindowPos(_interopWindowHandle, HWND_TOP, xPos, yPos, windowsWidth, windowsHeight, SWP_SHOWWINDOW));
```
Comment (carlos-zamora) [inline src/cascadia/WindowsTerminal/NonClientIslandWindow.cpp]: >HWND_TOP [](start = 57, length = 8)

Just double checking, you changed this from `HWND_BOTTOM` to `HWND_TOP`. Intentional?

File: src/cascadia/WindowsTerminal/NonClientIslandWindow.cpp [hunk 2]
```diff
@@ -113,8 +118,53 @@ void NonClientIslandWindow::OnSize(const UINT width, const UINT height)
...
+{
     if (_dragBar)
     {
+        // TODO:GH#1897 This is largely duplicated from OnSize, and we should do
+        // better than that.
+        const auto windowRect = GetWindowRect();
+        const auto width = windowRect.right - windowRect.left;
+        const auto height = windowRect.bottom - windowRect.top;
+
+        const auto scale = GetCurrentDpiScale();
```
Comment (carlos-zamora) [inline src/cascadia/WindowsTerminal/NonClientIslandWindow.cpp]: >const auto scale = GetCurrentDpiScale(); [](start = 8, length = 40)

I don't think this is actually used anywhere.

File: src/cascadia/WindowsTerminal/NonClientIslandWindow.cpp [hunk 3]
```diff
@@ -229,10 +277,13 @@ MARGINS NonClientIslandWindow::GetFrameMargins() const noexcept
 // - the HRESULT returned by DwmExtendFrameIntoClientArea.
 [[nodiscard]] HRESULT NonClientIslandWindow::_UpdateFrameMargins() const noexcept
 {
-    // Get the size of the borders we want to use. The sides and bottom will
-    // just be big enough for resizing, but the top will be as big as we need
-    // for the non-client content.
-    MARGINS margins = GetFrameMargins();
+    // Set frame margines with just a single pixel on the bottom. We don't
```
Comment (carlos-zamora) [inline src/cascadia/WindowsTerminal/NonClientIslandWindow.cpp]: >margines [](start = 17, length = 8)

nit: margins

File: src/cascadia/WindowsTerminal/NonClientIslandWindow.cpp [hunk 4]
```diff
@@ -386,61 +437,72 @@ RECT NonClientIslandWindow::GetMaxWindowRectInPixels(const RECT* const prcSugges
...
         break;
     }
+
     case WM_EXITSIZEMOVE:
     {
         ForceResize();
         break;
     }
-    case WM_NCACTIVATE:
-    case WM_NCPAINT:
```
Comment (carlos-zamora) [inline src/cascadia/WindowsTerminal/NonClientIslandWindow.cpp]: >case WM_NCPAINT: [](start = 4, length = 16)

What happens if we keep `WM_NCPAINT` here along with `WM_PAINT`? (completely new territory for me)

**Comment by DHowett-MSFT** [comment]
Bad news!
When maximized _with the button_, the hole isn't painted.
image

File: src/cascadia/WindowsTerminal/NonClientIslandWindow.cpp (hunk 1 above)
Comment (zadjii-msft) [reply src/cascadia/WindowsTerminal/NonClientIslandWindow.cpp]: It appears that change is actually unnecessary, so I'll revert it.

File: src/cascadia/WindowsTerminal/NonClientIslandWindow.cpp (hunk 4 above)
Comment (zadjii-msft) [reply src/cascadia/WindowsTerminal/NonClientIslandWindow.cpp]: If we handle `WM_NCPAINT` at all (by returning anything for a NCPAINT message), then the system assumes that we wanted to paint our nonclient area. Because it assumes we're doing _all_ the work, it does _nothing_. Unfortunately, the kernel is weird, so if the OS is doing nothing, the kernel still treats our window like a Windows xp/vista-like window, and the _kernel_ thinks that we have rounded top corners. The OS also won't paint our window shadow anymore, because it assumes we're doing that.

Basically everything is terrible if we try doing that.

**Comment by zadjii-msft** [comment]
@DHowett-MSFT For the record, I fixed the titlebar maximize thing in 9040940 (what a sha)

File: src/cascadia/WindowsTerminal/NonClientIslandWindow.cpp [hunk 5]
```diff
@@ -113,8 +118,53 @@ void NonClientIslandWindow::OnSize(const UINT width, const UINT height)
         _rootGrid.Arrange(finalRect);
     }
 
+    // I'm not sure that HWND_BOTTOM is any different than HWND_TOP for us.
```
Comment (miniksa) [inline src/cascadia/WindowsTerminal/NonClientIslandWindow.cpp]: I don't understand what this comment is supposed to mean.

Comment (zadjii-msft) [reply src/cascadia/WindowsTerminal/NonClientIslandWindow.cpp]: Oh mainly just that both those values do the same thing, which doesn't _really_ make sense, so there's not really a reason one was picked over the other
